compiler/qsc_parse/src/item.rs

```diff
@@ -44,12 +44,17 @@ pub(super) fn parse(s: &mut Scanner) -> Result<Box<Item>> {
     } else if let Some(callable) = opt(s, parse_callable_decl)? {
         Box::new(ItemKind::Callable(callable))
     } else {
-        return Err(Error(ErrorKind::Rule("item", s.peek().kind, s.peek().span)));
+        let p = s.peek();
+        if !doc.is_empty() {
+            s.push_error(Error(ErrorKind::FloatingDocComment(t.span)));
+            return Ok(Box::default());
```

Comment:
I agree with all of the above, thanks for the feedback.

Although you're right that `Item::default` isn't explicitly used anywhere, there may be instances of `unwrap_or_default` elsewhere, where it might be returning the default `Item` to prevent consuming tokens unnecessarily. I could be wrong though (have yet to check), and in this case an error recovery item sounds appealing. 

A guard in this case makes sense, since the error and the item are a pair.

Reply:
We have precedent for using `ItemKind::Err` for creating error nodes, but they follow a different pattern. See the `default` function in this file and how it's used. It should take a valid span (which would fix the `0-0` spans you're getting in the tests). See `default` in `stmt.rs` for a similar pattern for statements. I think you can return `Ok(default(s.span(lo)))` here to follow that pattern.

I kind of agree that creating error recovery nodes should always go together with calling `push_error`.. Currently this is ensured by only ever creating error nodes in `recover` (in `prim.rs`). This will be the only new place we're creating error nodes. I like the idea of extracting a helper method like @sezna suggested but I don't know how we would do it without looking really awkward, so I can take it or leave it.

Reply:
I agree with Mine. `ItemKind::Err` is a good pattern to use here, even if it's the first occurrence of that outside of the recovery code. We knew parser recovery was going to be along tail of improvements, so it's likely this won't be be the last time we need this pattern, and the refactoring Alex recommends would be a good improvement to introduce once we have a second instance to incorporate.

Reply:
Thanks for all the feedback, I should have some time near the end of the week to update this branch.

Reply:
Okay, I think it's sitting in a decent spot now. WRT the helper method, would that belong in `prim.rs`?

The current logic I have could be expressed as:
```rust
pub fn recovering_doc(s: &mut Scanner, span: Span) -> Box<Item> {
    let err_item = default(s.span(lo));
    s.push_error(Error(ErrorKind::FloatingDocComment(err_item.span)));
    err_item
}
```
Though it doesn't look much like the other recovering functions since it doesn't make use of the `token` function, nor `recover`. It could just be grouped in `item.rs` with `parse_doc`, since it really would just wrap the logic together, returning `default`.

Otherwise, I believe I've hammered out the rest of the requested changes. Let me know if there's anything else I can do, and thanks again for all the great feedback.

Reply:
I like the way you have it now. Perhaps we can keep an eye on it, and if more duplication crops up as we add more error recovery, we'll come up with a better design.